Doc/regions.md

```diff
@@ -0,0 +1,453 @@
+
+* Local objects can reach anything
+* Immutable objects can only reach other immutable objects.
+* Explicit objects can only reach explicit objects in the same region, immutable objects, or linear region entry points.
+* Explicit regions track the number of incoming references from local objects.
+
+New objects are always created as a local object.
+If Pyrona features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
+
+### Explicit regions
+
+Explicit regions are used to pass mutable state between concurrent units of execution (behaviours).
+For usability, we need local references into the region, that are not part of the region itself.
+This is precisely what local objects are for as they can refer into the region.
+
+The core challenge is allowing local objects to refer into an explicit region,
+while allowing the explicit region to be sent to another thread.
+To do this, we need to track the number of references from local objects to an explicit region, the "local reference count" (LRC).
+
+When we send an explicit region to another behaviour we need to ensure that the LRC is zero.
+To make this more usable, if the LRC is not zero, we can scan the current frame to find all references to the region, and remove them.
+If there are still references, then we can scan all objects (from the cycle detector DLL) to find a reference, and report that as an error.
+
+When a local object is promoted to an explicit region, we need to ensure that the LRC is correctly maintained.
+This involves adding new counts for reference to the newly added objects, and removing counts for edges that have become internal to the region.
+Consider the following example:
+
+```python
+x = {}
+#  x is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+y = {}
+#  y is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+x.f = y
+# x.f is a field of the dictionary object, which is local.
+# y is also local so no need to do anything in the write barrier.
+
+# Create a fresh region
+r = Region()
+# r is an explicit region
+r.f = y
+# r.f is a field of the region object, which is explicit.
+# y was local, so we need to promote y to the explicit region.
+# Before the assignment y has an RC of 2: One from the frame object, one from the object pointed to be x.
+# This adds two local RCs to the explicit region.
+y = None
+# This reduces the localRC on r, by one. 
+x = None
+# This causes the old value of x to be collected
+# This in turn reduces the localRC on r by one as it clears the f field that points into the explicit region.
+
+# We can now send the region r. It only has a single local RC r, and the operation will cause the 
+# local frame to clear that reference.
+send(r)
+# r is None at this point
```

Comment:
Just so I understand this new syntax: don't we actually want something like `cown = send(r)`? It isn't clear what send does here, and maybe it doesn't matter, but it seems like it does what `make_shareable()` does now?

Reply:
It is like `make_shareable`, or capturing something in the closure of a behaviour.  There are many different things that all have the same requirements from regions, so I was trying to be abstract.

Reply:
makes sense, once I got to the plan of work I understood that this was a placeholder of some kind to allow breaking up the work.